### ISSUE TITLE
Should we make utility classes immutable?

### ISSUE THREAD

**Issue by tedgoas-dialpad** [dialpad/dialtone#96]
We should make utility classes immutable by adding an `!important` to them so they can win specificity battles with legacy CSS in the our products? We should also convert all dialtone classes to using CSS and `@variables` instead of `.mixins` so that components can be overridden by utility classes.

**Make immutable:**
- [x] margin
- [x] padding
- [x] font size
- [x] font weight
- [x] style
- [x] text decoration
- [x] text transform
- [x] whitespace
- [x] word break
- [x] overflow wrap
- [ ] list style
- [x] line-height
- [x] color
- [x] background color
- [ ] background position
- [ ] border width
- [ ] border style
- [ ] border color
- [ ] border radius
- [x] flexbox
- [ ] width (+ min and max)
- [ ] height (+ min and max)
- [x] display
- [ ] cursor
- [ ] box sizing
- [ ] float
- [ ] vertical align
- [x] position
- [x] overflow
- [ ] opacity
- [ ] z-index
- [ ] outline
- [ ] box shadow
- [ ] transition

**Comment by tedgoas-dialpad** [comment]
@hynes-dialpad @dbecherdialpad I removed the `Question` tag because I don't think there's any question if we plan on using Dialtone classes directly in HTML. We need to add an `!important ` to utility classes because we have so many nested selectors in our codebase.

**Comment by dbecherdialpad** [comment]
One big downside of this is you cannot then override one utility class with another, such as:

```vue
      class="d-bgc-white"
      :class="{ 'd-bgc-purple-100': isSelected }"
```

Not sure if this works with the `h:` hover modifiers as well.

**Comment by tedgoas-dialpad** [comment]
This discussion started almost a year ago. Since then we've hired @dialpad-drew, so I defer to him on how we write CSS.

That said I'd argue there's still a strong case for having at least a handful of utility classes around to build and style basic layouts. WIth that in mind, here are a few ways to deal with your example above:

---

**1. Remove the class you want to override**
Initial state: `<div class="d-bgc-white">`
Selected state: `<div class="d-bgc-purple-100">`

So basically remove `d-bgc-white` so you don't have to override it.

---

**2. Don't use utility classes here.**
```
<div class="custom-thing"></div>

<style>
  .custom-thing {
    background-color: white;
    &.isSelected {
      background-color: var(--purple-100);
    }
  }
</style>
```

We don't need to do *everything* in utility classes, but we shouldn't need to write a custom classes just to add a background color and some spacing.

**Comment by dbecherdialpad** [comment]
To be clear my comment was about the `!important`, not the utility classes in general. In this case it was a quick prototype of something where the convenience of utility classes was nice when I just wanted to change one property. I could have done it by removing the white BG but that was slightly more work 😝 

My only comment on utility classes in general is that it's kind of annoying when you get 95% of the way with the utilities and then end up having to bail out and write CSS anyway. DT6 solves some of this (eg hover utilities, which _do_ work even with important, somehow), and some of it is just CSS that's complex enough it shouldn't use utilities anyway.
